opt: Add optimizer support for UPSERT and INSERT..ON CONFLICT statements

andy-kimball opened the pull request:
This commit adds support for three additional mutation operator cases:

  1. UPSERT statement
  2. INSERT..ON CONFLICT DO NOTHING
  3. INSERT..ON CONFLICT..DO UPDATE

Each of these has overlapping functionality with INSERT, UPDATE, and
with each other, but also have important differences. Similar to the
other mutation operators, the Upsert operator accepts a relational
input expression that projects columns that will be used as part of
executing the mutation:

  1. Columns containing existing values fetched from the target table
     and used to detect conflicts and to formulate the key/value update
     commands.

  2. Columns containing updated values to set when a conflict is
     detected, as specified by the user.

  3. Computed columns which will be updated when a conflict is detected
     and that are dependent on one or more updated columns.

For example, if this is the schema and INSERT..ON CONFLICT statement:

  CREATE TABLE abc (a INT PRIMARY KEY, b INT, c INT)
  INSERT INTO abc VALUES (1, 2) ON CONFLICT (a) DO UPDATE SET b=10

Then an input expression equivalent to this would be built:

  SELECT ins_a, ins_b, ins_c, fetch_a, fetch_b, fetch_c, 10 AS upd_b
  FROM (VALUES (1, 2, NULL)) AS ins(ins_a, ins_b, ins_c)
  LEFT OUTER JOIN abc AS fetch(fetch_a, fetch_b, fetch_c)
  ON ins_a = fetch_a

At runtime, the Upsert execution operator will test the fetch_a column.
If it is null, then there is no existing row, so the operator inserts a
new row using the (ins_a, ins_b, ins_c) columns. Otherwise, the operator
formulates an update using the (upd_b) column, along with any needed
fetch columns.

The CBO Upsert support is disabled by default behind a feature flag:
experimental_optimizer_updates.

andy-kimball committed:
sql: Remove tree.Datums return value from tableWriter.row method

This return value is not ever used by callers, so this commit removes it.

Release note: None

andy-kimball committed:
opt: Add Index.IsUnique and FindTableColumnByName catalog functions

The INSERT..ON CONFLICT statement needs to know if an index was declared
as UNIQUE by the user. This commit adds an IsUnique() method to the
cat.Index interface that exposes that information. It also adds a
cat.FindTableColumnByName method, which will also be used by upsert.

Release note: None

andy-kimball committed:
opt: Separate Insert/Update specific methods from mutationBuilder

Move "ForInsert" and "ForUpdate" methods from mutationBuilder to the
insert.go and update.go files. Leave shared utility methods in the
mutation_builder.go file. This sets the pattern for locating Upsert
and Delete specific files as well.

Release note: None

andy-kimball committed:
opt: Standardize on using "alias" as column name

Use the word "alias" rather than "label" more consistently, in order
to match the new Metadata terminology.

Release note: None

andy-kimball committed:
opt: Add Upsert support to the cost-based optimizer

This commit adds support for three additional mutation operator cases:

  1. UPSERT statement
  2. INSERT..ON CONFLICT DO NOTHING
  3. INSERT..ON CONFLICT..DO UPDATE

Each of these has overlapping functionality with INSERT, UPDATE, and
with each other, but also have important differences. Similar to the
other mutation operators, the Upsert operator accepts a relational
input expression that projects columns that will be used as part of
executing the mutation:

  1. Columns containing existing values fetched from the target table
     and used to detect conflicts and to formulate the key/value update
     commands.

  2. Columns containing updated values to set when a conflict is
     detected, as specified by the user.

  3. Computed columns which will be updated when a conflict is detected
     and that are dependent on one or more updated columns.

For example, if this is the schema and INSERT..ON CONFLICT statement:

  CREATE TABLE abc (a INT PRIMARY KEY, b INT, c INT)
  INSERT INTO abc VALUES (1, 2) ON CONFLICT (a) DO UPDATE SET b=10

Then an input expression equivalent to this would be built:

  SELECT ins_a, ins_b, ins_c, fetch_a, fetch_b, fetch_c, 10 AS upd_b
  FROM (VALUES (1, 2, NULL)) AS ins(ins_a, ins_b, ins_c)
  LEFT OUTER JOIN abc AS fetch(fetch_a, fetch_b, fetch_c)
  ON ins_a = fetch_a

At runtime, the Upsert execution operator will test the fetch_a column.
If it is null, then there is no existing row, so the operator inserts a
new row using the (ins_a, ins_b, ins_c) columns. Otherwise, the operator
formulates an update using the (upd_b) column, along with any needed
fetch columns.

The CBO Upsert support is disabled by default behind a feature flag:
experimental_optimizer_updates.

Release note (sql change): The new experimental_optimizer_updates feature
flag controls whether updates and upserts are planned by the cost-based
optimizer rather than the heuristic planner.

cockroach-teamcity commented:
<!-- Reviewable:start -->
This change is [<img src="https://reviewable.io/review_button.png" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/cockroachdb/cockroach/33339)
<!-- Reviewable:end -->

knz reviewed (commented):
First, second and last commit LGTM modulo some nits and a missing release note.

There is a change in the left joiner exec code in the last commit which I do not understand - does it belong there? It probably merits explanation in the commit message.

Also I'd like to understand (preferably from the commit/PR description) what your view is on the transitory period where the two implementations of UPSERT have different behavior:
- different conflict handling
- different fast path

Which one is going to be the default? If the CBO becomes the default this needs to be communicated amply to beta users. If not we need to have a path to exercise it thoroughly.

> Reviewed 10 of 10 files at r1, 8 of 8 files at r2, 26 of 27 files at r6.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/33339#-:-LUVM1ve7kYngCO57bpl:byfgwfb)* status: :shipit: complete! 0 of 0 LGTMs obtained

___
*[pkg/sql/plan.go, line 369 at r6](https://reviewable.io/reviews/cockroachdb/cockroach/33339#-LUVICLX6ugM8hDTSNOg:-LUVICLYCQAPbc_V7F0B:b-5g5a3f) ([raw file](https://github.com/cockroachdb/cockroach/blob/70e37d031d499630d757b1ef85a284bebb952e11/pkg/sql/plan.go#L369)):*
> ```Go
> 	p.curPlan = planTop{AST: stmt.AST}
> 
> 	//s := stmt.String()
> ```

nit: stray comments
___
*[pkg/sql/tablewriter_upsert_opt.go, line 214 at r6](https://reviewable.io/reviews/cockroachdb/cockroach/33339#-LUVIZXi5KcgtLekFiIn:-LUVIZXi5KcgtLekFiIo:bn1vdud) ([raw file](https://github.com/cockroachdb/cockroach/blob/70e37d031d499630d757b1ef85a284bebb952e11/pkg/sql/tablewriter_upsert_opt.go#L214)):*
> ```Go
> 			rowIndex, ok = tu.ru.FetchColIDtoRowIndex[colID]
> 			if !ok {
> 				panic("no existing value is available for column")
> ```

`return pgerror.NewAssertionErrorf(...)`

(this will go to telemetry where we can pick it up without bothering users with dead nodes)
___
*[pkg/sql/logictest/testdata/logic_test/check_constraints, line 157 at r6](https://reviewable.io/reviews/cockroachdb/cockroach/33339#-LUVIyKQEFPqnkP1Z4X6:-LUVIyKQEFPqnkP1Z4X7:bncop0m) ([raw file](https://github.com/cockroachdb/cockroach/blob/70e37d031d499630d757b1ef85a284bebb952e11/pkg/sql/logictest/testdata/logic_test/check_constraints#L157)):*
> ```
> # only checks constraints if an insert or update actually occurs. In this case,
> # the DO NOTHING clause skips the update, so there is no need to check the
> # constraint.
> ```

1) Can you check whether pg does the same or different

2) In either case, document the behavior change in a release note.
___
*[pkg/sql/logictest/testdata/logic_test/computed, line 734 at r6](https://reviewable.io/reviews/cockroachdb/cockroach/33339#-LUVJDKdE3HSD7qHkQBt:-LUVJDKdE3HSD7qHkQBu:b-hrnldb) ([raw file](https://github.com/cockroachdb/cockroach/blob/70e37d031d499630d757b1ef85a284bebb952e11/pkg/sql/logictest/testdata/logic_test/computed#L734)):*
> ```
> 
> # Upsert -> update
> # NOTE: The CBO cannot show the name of the computed column in the error message
> ```

Please remove the explanatory comment from here and instead add it to #33345.
___
*[pkg/sql/logictest/testdata/logic_test/privileges_table, line 245 at r6](https://reviewable.io/reviews/cockroachdb/cockroach/33339#-LUVK7Eq4wiwrOSA6gkq:-LUVK7Eq4wiwrOSA6gkr:b-eksqmn) ([raw file](https://github.com/cockroachdb/cockroach/blob/70e37d031d499630d757b1ef85a284bebb952e11/pkg/sql/logictest/testdata/logic_test/privileges_table#L245)):*
> ```
> # Grant SELECT and INSERT privilege.
> # TODO(andyk): The SELECT privilege is required by the optimizer to fetch
> # existing columns. This matches PG behavior, but differs from planner behavior.
> ```

This explanatory comment is not needed because we'll fix #33346.
___
*[pkg/sql/logictest/testdata/logic_test/txn, line 942 at r6](https://reviewable.io/reviews/cockroachdb/cockroach/33339#-LUVKiHe0entiiTLwpF_:-LUVKiHe0entiiTLwpFa:bhywy5t) ([raw file](https://github.com/cockroachdb/cockroach/blob/70e37d031d499630d757b1ef85a284bebb952e11/pkg/sql/logictest/testdata/logic_test/txn#L942)):*
> ```
> UPDATE kv SET v = 'foo'
> 
> # The planner uses INSERT, the optimizer uses UPSERT; accept both for now.
> ```

s/planner/heuristic planner/
___
*[pkg/sql/logictest/testdata/logic_test/upsert_opt, line 1 at r6](https://reviewable.io/reviews/cockroachdb/cockroach/33339#-LUVKsG693f9DkuFNLM8:-LUVKsG693f9DkuFNLM9:buvc05h) ([raw file](https://github.com/cockroachdb/cockroach/blob/70e37d031d499630d757b1ef85a284bebb952e11/pkg/sql/logictest/testdata/logic_test/upsert_opt#L1)):*
> ```
> # LogicTest: local-opt fakedist-opt
> ```

Could you paste a diff of the two files in the commit message, so that our future selves can satisfy themselves nothing important was lost in the copy/paste.
___
*[pkg/sql/opt/exec/execbuilder/relational_builder.go, line 1198 at r6](https://reviewable.io/reviews/cockroachdb/cockroach/33339#-LUVLNJi5WP2NgMFkEBb:-LUVLNJi5WP2NgMFkEBc:b-edb3pp) ([raw file](https://github.com/cockroachdb/cockroach/blob/70e37d031d499630d757b1ef85a284bebb952e11/pkg/sql/opt/exec/execbuilder/relational_builder.go#L1198)):*
> ```Go
> 	//
> 	// TODO(andyk): Using ensureColumns here can result in an extra Render.
> 	// Upgrade execution engine to not require this.
> ```

Can you spell out the required change? Better yet spell it out in a linked issue in the TODO text.


<!-- Sent from Reviewable.io -->

rytaft reviewed (changes requested):
Nice work! Mostly minor comments below...

> Reviewed 10 of 10 files at r1, 8 of 8 files at r2, 3 of 3 files at r3, 9 of 9 files at r4, 16 of 16 files at r5, 27 of 27 files at r6.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/33339#-:-LVAlu-Z2RMqhq1QWl2D:b-1ey85g)* status: :shipit: complete! 0 of 0 LGTMs obtained

___
*[pkg/sql/opt_catalog.go, line 509 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/33339#-LUr-q8L0m2l27W_gN5c:-LUr-q8M2TfP6Doj4IHY:bzbfkbx) ([raw file](https://github.com/cockroachdb/cockroach/blob/7ab0493642a1e1b8c268b8ea7a7427d3ed3f3354/pkg/sql/opt_catalog.go#L509)):*
> ```Go
> }
> 
> // IsIndex is part of the cat.Index interface.
> ```

IsIndex -> IsUnique
___
*[pkg/sql/logictest/testdata/logic_test/upsert_opt, line 1 at r6](https://reviewable.io/reviews/cockroachdb/cockroach/33339#-LUVKsG693f9DkuFNLM8:-LVAi2SuCYEenVo3vTIr:b-ckmmf0) ([raw file](https://github.com/cockroachdb/cockroach/blob/70e37d031d499630d757b1ef85a284bebb952e11/pkg/sql/logictest/testdata/logic_test/upsert_opt#L1)):*
<details><summary><i>Previously, knz (kena) wrote…</i></summary><blockquote>

Could you paste a diff of the two files in the commit message, so that our future selves can satisfy themselves nothing important was lost in the copy/paste.
</blockquote></details>

It would also help to add a comment in this file explaining what's different
___
*[pkg/sql/opt/ops/statement.opt, line 127 at r5](https://reviewable.io/reviews/cockroachdb/cockroach/33339#-LUr8vrA8BePFzz0yXW8:-LUr8vrA8BePFzz0yXW9:bye143t) ([raw file](https://github.com/cockroachdb/cockroach/blob/2344d2e2148b16c6d4e237929edd8620ae27c1ea/pkg/sql/opt/ops/statement.opt#L127)):*
> ```
> # into a target table. If a conflicting row already exists, then Upsert will
> # instead update the existing row. The Upsert operator uses insertion, existing,
> # and new values constructed by the input to insert rows, update indexes, and to
> ```

I think this sentence is trying to do a bit too much... it would help to break this up and explain the difference between an "insertion" value and a "new" value.
___
*[pkg/sql/opt/optbuilder/insert.go, line 32 at r5](https://reviewable.io/reviews/cockroachdb/cockroach/33339#-LUr9nbzE-9lER1sn3mL:-LUr9nc-AQoGzep1rARV:briscs) ([raw file](https://github.com/cockroachdb/cockroach/blob/2344d2e2148b16c6d4e237929edd8620ae27c1ea/pkg/sql/opt/optbuilder/insert.go#L32)):*
> ```Go
> 
> // excludedTableName is the name of a special Upsert data source containing
> // columns that could not be inserted due to a conflict with an existing row:
> ```

do you mean "rows that could not be inserted..."?
___
*[pkg/sql/opt/optbuilder/insert.go, line 474 at r5](https://reviewable.io/reviews/cockroachdb/cockroach/33339#-LUrEXJkBUeKHrY8kVhh:-LUrEXJl8vRv2raRyJiJ:b-g485qf) ([raw file](https://github.com/cockroachdb/cockroach/blob/2344d2e2148b16c6d4e237929edd8620ae27c1ea/pkg/sql/opt/optbuilder/insert.go#L474)):*
> ```Go
> 	//   2. Assign name to each column
> 	//   3. Add id of each column to the insertColList
> 	mb.insertColList = make(opt.ColList, cap(mb.targetColList))
> ```

Wasn't this already done at the top of the function?
___
*[pkg/sql/opt/optbuilder/insert.go, line 533 at r5](https://reviewable.io/reviews/cockroachdb/cockroach/33339#-LVARWwPArhsWUeKboPn:-LVARWwPArhsWUeKboPo:b-kx5svp) ([raw file](https://github.com/cockroachdb/cockroach/blob/2344d2e2148b16c6d4e237929edd8620ae27c1ea/pkg/sql/opt/optbuilder/insert.go#L533)):*
> ```Go
> // filter that discards rows that have a conflict (by checking a not-null table
> // column to see if it was null-extended by the left join). See the comment
> // header for buildInsert for an example.
> ```

It would be good to specify `Builder.buildInsert` to differentiate from `mutationBuilder.buildInsert`
___
*[pkg/sql/opt/optbuilder/insert.go, line 538 at r5](https://reviewable.io/reviews/cockroachdb/cockroach/33339#-LVANZ9k--UzhgVRyhv7:-LVANZ9k--UzhgVRyhv8:b-u6owpx) ([raw file](https://github.com/cockroachdb/cockroach/blob/2344d2e2148b16c6d4e237929edd8620ae27c1ea/pkg/sql/opt/optbuilder/insert.go#L538)):*
> ```Go
> 	var conflictIndex cat.Index
> 	if len(onConflict.Columns) != 0 {
> 		// Check that the ON CONFLICT columns reference at most one target row.
> ```

This is a bit confusing - how can columns reference at most one row? 
___
*[pkg/sql/opt/optbuilder/insert.go, line 556 at r5](https://reviewable.io/reviews/cockroachdb/cockroach/33339#-LVANhp83MVwPOeNQfKx:-LVANhp83MVwPOeNQfKy:b-d9rgtt) ([raw file](https://github.com/cockroachdb/cockroach/blob/2344d2e2148b16c6d4e237929edd8620ae27c1ea/pkg/sql/opt/optbuilder/insert.go#L556)):*
> ```Go
> 
> 		// If conflict columns were explicitly specified, then only check for a
> 		// conflict one a single index. Otherwise, check on all indexes.
> ```

one -> on
___
*[pkg/sql/opt/optbuilder/insert.go, line 580 at r5](https://reviewable.io/reviews/cockroachdb/cockroach/33339#-LVASYdQ1tMsqNdbO1r_:-LVASYdQ1tMsqNdbO1ra:b9tloa8) ([raw file](https://github.com/cockroachdb/cockroach/blob/2344d2e2148b16c6d4e237929edd8620ae27c1ea/pkg/sql/opt/optbuilder/insert.go#L580)):*
> ```Go
> 		// Remember the column ID of a scan column that is not null. This will be
> 		// used to detect whether a conflict was detected for a row.
> 		notNullColID := scanScope.cols[findNotNullIndexCol(index)].id
> ```

would be good to explain why this always exists and no canary col is necessary
___
*[pkg/sql/opt/optbuilder/insert.go, line 602 at r5](https://reviewable.io/reviews/cockroachdb/cockroach/33339#-LVAQaYl5gLvRqM47TZg:-LVAQaYl5gLvRqM47TZh:bdqmhol) ([raw file](https://github.com/cockroachdb/cockroach/blob/2344d2e2148b16c6d4e237929edd8620ae27c1ea/pkg/sql/opt/optbuilder/insert.go#L602)):*
> ```Go
> 		mb.outScope.expr = mb.b.factory.ConstructProject(
> 			mb.b.factory.ConstructSelect(
> 				mb.b.factory.ConstructLeftJoin(
> ```

Why do you need one left join for each unique index? It seems like you could get away with a single left join, with the join conditions for different indexes OR-ed together.
___
*[pkg/sql/opt/optbuilder/insert.go, line 644 at r5](https://reviewable.io/reviews/cockroachdb/cockroach/33339#-LVATAssBR0TWG4XZFoj:-LVATAssBR0TWG4XZFok:b2dtfm8) ([raw file](https://github.com/cockroachdb/cockroach/blob/2344d2e2148b16c6d4e237929edd8620ae27c1ea/pkg/sql/opt/optbuilder/insert.go#L644)):*
> ```Go
> 	}
> 
> 	// Build the right side of the left outer join. Include mutation columns
> ```

Seems like a lot of this code is duplicated from the function above. Is there any way to consolidate?
___
*[pkg/sql/opt/optbuilder/insert.go, line 735 at r5](https://reviewable.io/reviews/cockroachdb/cockroach/33339#-LVATk8h3UWLHLRSF6Kq:-LVATk8h3UWLHLRSF6Kr:buxeotf) ([raw file](https://github.com/cockroachdb/cockroach/blob/2344d2e2148b16c6d4e237929edd8620ae27c1ea/pkg/sql/opt/optbuilder/insert.go#L735)):*
> ```Go
> 	if len(insertCols) != 0 {
> 		for _, name := range insertCols {
> 			if ord := cat.FindTableColumnByName(mb.tab, name); ord != -1 {
> ```

do you need to throw an error if the column is not found?
___
*[pkg/sql/opt/optbuilder/insert.go, line 796 at r5](https://reviewable.io/reviews/cockroachdb/cockroach/33339#-LVAV-8o9X9k_tiD6Pc8:-LVAV-8p4IMHnCMCejaa:b-ekhgrr) ([raw file](https://github.com/cockroachdb/cockroach/blob/2344d2e2148b16c6d4e237929edd8620ae27c1ea/pkg/sql/opt/optbuilder/insert.go#L796)):*
> ```Go
> 	}
> 	panic(builderError{errors.New(
> 		"there is no unique or exclusion constraint matching the ON CONFLICT specification")})
> ```

Is there a pgcode for this error (since you said Postgres also requires this)?
___
*[pkg/sql/opt/optbuilder/testdata/upsert, line 258 at r5](https://reviewable.io/reviews/cockroachdb/cockroach/33339#-LVAZSS74ejggaftyR3K:-LVAZSS88W1ytu9KXPGd:b4ot13g) ([raw file](https://github.com/cockroachdb/cockroach/blob/2344d2e2148b16c6d4e237929edd8620ae27c1ea/pkg/sql/opt/optbuilder/testdata/upsert#L258)):*
> ```
>                 └── const: 1 [type=int]
> 
> # Use RETURNING INSERT..ON CONFLICT as a FROM clause.
> ```

This example doesn't have ON CONFLICT


<!-- Sent from Reviewable.io -->

andy-kimball committed:
opt: Add execution support for Upsert operator

Build the upsertNode for the CBO Upsert operator. Also, create a new
upsert tableWriter that is custom built for the CBO. The CBO can use
a much simpler upserter because it incorporates conflict detection,
update and computed column evaluation, and other upsert operations
into the input query, rather than requiring the upserter to do it.

Release note (sql change): when the experimental_optimizer_updates
feature flag is set, check constraints are not checked for rows
skipped by the INSERT..DO NOTHING clause.

andy-kimball reviewed (commented):
Regarding the default, I already have a note in the commit message:
```
The CBO Upsert support is disabled by default behind a feature flag:
experimental_optimizer_updates.
```
I will only make it the default once the equivalent fast path(s) are in place. In addition, I believe your separate PR has changed the conflict handling to be equivalent, so that shouldn't be a problem either.

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/33339#-:-LVELCqo160KJXz-abk9:b-44jpi7)* status: :shipit: complete! 0 of 0 LGTMs obtained

___
*[pkg/sql/opt_catalog.go, line 509 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/33339#-LUr-q8L0m2l27W_gN5c:-LVESe35BllzVQl8Uumf:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/7ab0493642a1e1b8c268b8ea7a7427d3ed3f3354/pkg/sql/opt_catalog.go#L509)):*
<details><summary><i>Previously, rytaft wrote…</i></summary><blockquote>

IsIndex -> IsUnique
</blockquote></details>

Done.
___
*[pkg/sql/tablewriter_upsert_opt.go, line 214 at r6](https://reviewable.io/reviews/cockroachdb/cockroach/33339#-LUVIZXi5KcgtLekFiIn:-LVESxDRBYqPajyDQ1Kt:b97nyl9) ([raw file](https://github.com/cockroachdb/cockroach/blob/70e37d031d499630d757b1ef85a284bebb952e11/pkg/sql/tablewriter_upsert_opt.go#L214)):*
<details><summary><i>Previously, knz (kena) wrote…</i></summary><blockquote>

`return pgerror.NewAssertionErrorf(...)`

(this will go to telemetry where we can pick it up without bothering users with dead nodes)
</blockquote></details>

Ah, nice. Done.
___
*[pkg/sql/logictest/testdata/logic_test/check_constraints, line 157 at r6](https://reviewable.io/reviews/cockroachdb/cockroach/33339#-LUVIyKQEFPqnkP1Z4X6:-LVET8nTCOnfUKZ0Lm0i:b-70vqqn) ([raw file](https://github.com/cockroachdb/cockroach/blob/70e37d031d499630d757b1ef85a284bebb952e11/pkg/sql/logictest/testdata/logic_test/check_constraints#L157)):*
<details><summary><i>Previously, knz (kena) wrote…</i></summary><blockquote>

1) Can you check whether pg does the same or different

2) In either case, document the behavior change in a release note.

</blockquote></details>

PG behaves the same way that the HP does. Note that I expect this difference in behavior to be temporary, until I'm able to incorporate check constraints into the CBO. I added a bit to the comment:
```
# TODO(andyk): The CBO behaves differently than the HP and PG in this case. It
# only checks constraints if an insert or update actually occurs. In this case,
# the DO NOTHING clause skips the update, so there is no need to check the
# constraint. Once the CBO incorporates check constraints, this difference in
# behavior should go away.
```
Since the CBO is off-by-default, there's no need for a release note.
___
*[pkg/sql/logictest/testdata/logic_test/computed, line 734 at r6](https://reviewable.io/reviews/cockroachdb/cockroach/33339#-LUVJDKdE3HSD7qHkQBt:-LVEZZsM3hbEq9Ittkyj:bd40vu0) ([raw file](https://github.com/cockroachdb/cockroach/blob/70e37d031d499630d757b1ef85a284bebb952e11/pkg/sql/logictest/testdata/logic_test/computed#L734)):*
<details><summary><i>Previously, knz (kena) wrote…</i></summary><blockquote>

Please remove the explanatory comment from here and instead add it to #33345.
</blockquote></details>

How would that issue address this difference in behavior between the CBO and HP? The CBO implements computed columns as projections, and they get evaluated as part of the larger SQL expression. Adding context fields to the pgerror won't have any effect on how computed columns work in the CBO.
___
*[pkg/sql/logictest/testdata/logic_test/txn, line 942 at r6](https://reviewable.io/reviews/cockroachdb/cockroach/33339#-LUVKiHe0entiiTLwpF_:-LVEYVxC-vAaDdy2pGcx:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/70e37d031d499630d757b1ef85a284bebb952e11/pkg/sql/logictest/testdata/logic_test/txn#L942)):*
<details><summary><i>Previously, knz (kena) wrote…</i></summary><blockquote>

s/planner/heuristic planner/
</blockquote></details>

Done.
___
*[pkg/sql/opt/exec/execbuilder/relational_builder.go, line 1198 at r6](https://reviewable.io/reviews/cockroachdb/cockroach/33339#-LUVLNJi5WP2NgMFkEBb:-LVO57KY8YQLDBC6t9k7:b-s6bxgw) ([raw file](https://github.com/cockroachdb/cockroach/blob/70e37d031d499630d757b1ef85a284bebb952e11/pkg/sql/opt/exec/execbuilder/relational_builder.go#L1198)):*
<details><summary><i>Previously, knz (kena) wrote…</i></summary><blockquote>

Can you spell out the required change? Better yet spell it out in a linked issue in the TODO text.
</blockquote></details>

I don't have more detail on the changes that will need to be made. The TODO is there to remind us to delve more deeply into this when we have time.
___
*[pkg/sql/opt/ops/statement.opt, line 127 at r5](https://reviewable.io/reviews/cockroachdb/cockroach/33339#-LUr8vrA8BePFzz0yXW8:-LVO68_g5auJ-a_WQxtt:brt2mq3) ([raw file](https://github.com/cockroachdb/cockroach/blob/2344d2e2148b16c6d4e237929edd8620ae27c1ea/pkg/sql/opt/ops/statement.opt#L127)):*
<details><summary><i>Previously, rytaft wrote…</i></summary><blockquote>

I think this sentence is trying to do a bit too much... it would help to break this up and explain the difference between an "insertion" value and a "new" value.
</blockquote></details>

I got rid of the sentence altogether. There is a ton of other information already in `MutationPrivate` comments, as well as `Builder.buildInsert` comments.
___
*[pkg/sql/opt/optbuilder/insert.go, line 32 at r5](https://reviewable.io/reviews/cockroachdb/cockroach/33339#-LUr9nbzE-9lER1sn3mL:-LVO7B505WAgb_Kivsgg:b-qxhxn2) ([raw file](https://github.com/cockroachdb/cockroach/blob/2344d2e2148b16c6d4e237929edd8620ae27c1ea/pkg/sql/opt/optbuilder/insert.go#L32)):*
<details><summary><i>Previously, rytaft wrote…</i></summary><blockquote>

do you mean "rows that could not be inserted..."?
</blockquote></details>

I rewrote the sentence.
___
*[pkg/sql/opt/optbuilder/insert.go, line 474 at r5](https://reviewable.io/reviews/cockroachdb/cockroach/33339#-LUrEXJkBUeKHrY8kVhh:-LVO7J0NDKyN4LOONBtv:bobclp9) ([raw file](https://github.com/cockroachdb/cockroach/blob/2344d2e2148b16c6d4e237929edd8620ae27c1ea/pkg/sql/opt/optbuilder/insert.go#L474)):*
<details><summary><i>Previously, rytaft wrote…</i></summary><blockquote>

Wasn't this already done at the top of the function?
</blockquote></details>

Good catch. Removed.
___
*[pkg/sql/opt/optbuilder/insert.go, line 533 at r5](https://reviewable.io/reviews/cockroachdb/cockroach/33339#-LVARWwPArhsWUeKboPn:-LVO7Q2e-DGT6D3_7g9A:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/2344d2e2148b16c6d4e237929edd8620ae27c1ea/pkg/sql/opt/optbuilder/insert.go#L533)):*
<details><summary><i>Previously, rytaft wrote…</i></summary><blockquote>

It would be good to specify `Builder.buildInsert` to differentiate from `mutationBuilder.buildInsert`
</blockquote></details>

Done.
___
*[pkg/sql/opt/optbuilder/insert.go, line 538 at r5](https://reviewable.io/reviews/cockroachdb/cockroach/33339#-LVANZ9k--UzhgVRyhv7:-LVO7lPlAfk9d-XrtpsI:b6a5ulc) ([raw file](https://github.com/cockroachdb/cockroach/blob/2344d2e2148b16c6d4e237929edd8620ae27c1ea/pkg/sql/opt/optbuilder/insert.go#L538)):*
<details><summary><i>Previously, rytaft wrote…</i></summary><blockquote>

This is a bit confusing - how can columns reference at most one row? 
</blockquote></details>

Added a bit of extra text to clarify.
___
*[pkg/sql/opt/optbuilder/insert.go, line 556 at r5](https://reviewable.io/reviews/cockroachdb/cockroach/33339#-LVANhp83MVwPOeNQfKx:-LVO7uSB-EnORMh_-eOO:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/2344d2e2148b16c6d4e237929edd8620ae27c1ea/pkg/sql/opt/optbuilder/insert.go#L556)):*
<details><summary><i>Previously, rytaft wrote…</i></summary><blockquote>

one -> on
</blockquote></details>

Done.
___
*[pkg/sql/opt/optbuilder/insert.go, line 580 at r5](https://reviewable.io/reviews/cockroachdb/cockroach/33339#-LVASYdQ1tMsqNdbO1r_:-LVO8RXf1H-ZBz0u-X_g:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/2344d2e2148b16c6d4e237929edd8620ae27c1ea/pkg/sql/opt/optbuilder/insert.go#L580)):*
<details><summary><i>Previously, rytaft wrote…</i></summary><blockquote>

would be good to explain why this always exists and no canary col is necessary
</blockquote></details>

Done.
___
*[pkg/sql/opt/optbuilder/insert.go, line 602 at r5](https://reviewable.io/reviews/cockroachdb/cockroach/33339#-LVAQaYl5gLvRqM47TZg:-LVO8TL9EMaS2S21m3LT:b-gnk5ul) ([raw file](https://github.com/cockroachdb/cockroach/blob/2344d2e2148b16c6d4e237929edd8620ae27c1ea/pkg/sql/opt/optbuilder/insert.go#L602)):*
<details><summary><i>Previously, rytaft wrote…</i></summary><blockquote>

Why do you need one left join for each unique index? It seems like you could get away with a single left join, with the join conditions for different indexes OR-ed together.
</blockquote></details>

It's possible to create a big predicate that does all the analysis, but then we'd select a bad plan for it. We'd end up just doing a primary index scan, and evaluate the big predicate on each row. Instead, we need to do a series of lookup joins (that's what the current Upsert code does).
___
*[pkg/sql/opt/optbuilder/insert.go, line 644 at r5](https://reviewable.io/reviews/cockroachdb/cockroach/33339#-LVATAssBR0TWG4XZFoj:-LVO9WEf8ZH011PfggDn:bfqo7kr) ([raw file](https://github.com/cockroachdb/cockroach/blob/2344d2e2148b16c6d4e237929edd8620ae27c1ea/pkg/sql/opt/optbuilder/insert.go#L644)):*
<details><summary><i>Previously, rytaft wrote…</i></summary><blockquote>

Seems like a lot of this code is duplicated from the function above. Is there any way to consolidate?
</blockquote></details>

There are definitely similarities, but there are also important differences that make it painful to consolidate. This is what I arrived at after trying out various ways of consolidating. I don't know how to make it better without making it harder to understand.
___
*[pkg/sql/opt/optbuilder/insert.go, line 735 at r5](https://reviewable.io/reviews/cockroachdb/cockroach/33339#-LVATk8h3UWLHLRSF6Kq:-LVOHoOE7RoBjinZXhrY:bkdr9s4) ([raw file](https://github.com/cockroachdb/cockroach/blob/2344d2e2148b16c6d4e237929edd8620ae27c1ea/pkg/sql/opt/optbuilder/insert.go#L735)):*
<details><summary><i>Previously, rytaft wrote…</i></summary><blockquote>

do you need to throw an error if the column is not found?
</blockquote></details>

They should always exist, because the existence of `insertCols` has already been checked. I changed the code to get rid of the unnecessary `if` and added a comment.
___
*[pkg/sql/opt/optbuilder/insert.go, line 796 at r5](https://reviewable.io/reviews/cockroachdb/cockroach/33339#-LVAV-8o9X9k_tiD6Pc8:-LVOIFUq940D7skMwu9L:bwks1l8) ([raw file](https://github.com/cockroachdb/cockroach/blob/2344d2e2148b16c6d4e237929edd8620ae27c1ea/pkg/sql/opt/optbuilder/insert.go#L796)):*
<details><summary><i>Previously, rytaft wrote…</i></summary><blockquote>

Is there a pgcode for this error (since you said Postgres also requires this)?
</blockquote></details>

Possibly, but I'm trying to match the behavior of the existing code as closely as possible, and it doesn't use a PG error code.
___
*[pkg/sql/opt/optbuilder/testdata/upsert, line 258 at r5](https://reviewable.io/reviews/cockroachdb/cockroach/33339#-LVAZSS74ejggaftyR3K:-LVONRmv-CkQBf0igyYL:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/2344d2e2148b16c6d4e237929edd8620ae27c1ea/pkg/sql/opt/optbuilder/testdata/upsert#L258)):*
<details><summary><i>Previously, rytaft wrote…</i></summary><blockquote>

This example doesn't have ON CONFLICT
</blockquote></details>

Done.
___
*[pkg/sql/logictest/testdata/logic_test/privileges_table, line 245 at r6](https://reviewable.io/reviews/cockroachdb/cockroach/33339#-LUVK7Eq4wiwrOSA6gkq:-LVEZTQIDDPhMlNEjb88:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/70e37d031d499630d757b1ef85a284bebb952e11/pkg/sql/logictest/testdata/logic_test/privileges_table#L245)):*
<details><summary><i>Previously, knz (kena) wrote…</i></summary><blockquote>

This explanatory comment is not needed because we'll fix #33346.
</blockquote></details>

Done.
___
*[pkg/sql/logictest/testdata/logic_test/upsert_opt, line 1 at r6](https://reviewable.io/reviews/cockroachdb/cockroach/33339#-LUVKsG693f9DkuFNLM8:-LVO2k0R3zpOtM5-oHYz:b-otpg36) ([raw file](https://github.com/cockroachdb/cockroach/blob/70e37d031d499630d757b1ef85a284bebb952e11/pkg/sql/logictest/testdata/logic_test/upsert_opt#L1)):*
<details><summary><i>Previously, rytaft wrote…</i></summary><blockquote>

It would also help to add a comment in this file explaining what's different
</blockquote></details>

The main reason for adding this file was due to differences in conflict handling between the CBO and HP, specifically in the case where the same row is inserted/updated multiple times in the same statement. However, after discussion, Raphael and I have agreed that we should not make any semantic guarantees in that case, and should instead require users to use `DISTINCT ON` to remove any duplicates:

  https://github.com/cockroachdb/docs/pull/4215

Therefore, I've updated the `upsert` logictest to reflect this decision, and I've removed `upsert_opt` altogether.


<!-- Sent from Reviewable.io -->

knz reviewed (commented):
I think a release note should also mention the existence of this flag (I don't recall seeing such a release note earlier in the first PR where you added it).

> Reviewed 3 of 62 files at r7, 3 of 25 files at r12.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/33339#-:-LVOdYV49Vf9iUxNkV3U:blm1a9o)* status: :shipit: complete! 0 of 0 LGTMs obtained

___
*[pkg/sql/logictest/testdata/logic_test/check_constraints, line 157 at r6](https://reviewable.io/reviews/cockroachdb/cockroach/33339#-LUVIyKQEFPqnkP1Z4X6:-LVOcRNT1zi2W2fMVWt-:bb4s9wc) ([raw file](https://github.com/cockroachdb/cockroach/blob/70e37d031d499630d757b1ef85a284bebb952e11/pkg/sql/logictest/testdata/logic_test/check_constraints#L157)):*
<details><summary><i>Previously, andy-kimball (Andy Kimball) wrote…</i></summary><blockquote>

PG behaves the same way that the HP does. Note that I expect this difference in behavior to be temporary, until I'm able to incorporate check constraints into the CBO. I added a bit to the comment:
```
# TODO(andyk): The CBO behaves differently than the HP and PG in this case. It
# only checks constraints if an insert or update actually occurs. In this case,
# the DO NOTHING clause skips the update, so there is no need to check the
# constraint. Once the CBO incorporates check constraints, this difference in
# behavior should go away.
```
Since the CBO is off-by-default, there's no need for a release note.
</blockquote></details>

Yeah so the situation with release notes is that if you don't write something about this now, nobody will notice the change in behavior later when we switch the default. We're not creating release notes out of thin air about all the behavior that becomes different when a flag is switched. Someone has to write it, and I don't see any record of this behavior change anywhere for the future release note writer to pick up.

My recommendation would be for a release note that says "when the experimental flag is set, the behavior is ..."
___
*[pkg/sql/logictest/testdata/logic_test/computed, line 734 at r6](https://reviewable.io/reviews/cockroachdb/cockroach/33339#-LUVJDKdE3HSD7qHkQBt:-LVOcraCBRBOpceSHYBw:b-jy3wm9) ([raw file](https://github.com/cockroachdb/cockroach/blob/70e37d031d499630d757b1ef85a284bebb952e11/pkg/sql/logictest/testdata/logic_test/computed#L734)):*
<details><summary><i>Previously, andy-kimball (Andy Kimball) wrote…</i></summary><blockquote>

How would that issue address this difference in behavior between the CBO and HP? The CBO implements computed columns as projections, and they get evaluated as part of the larger SQL expression. Adding context fields to the pgerror won't have any effect on how computed columns work in the CBO.
</blockquote></details>

1) I read your response to my review comment as an additional requirement on how we're going to produce meaningful contextual strings. It's a non-trivial problem but it's solvable.

2) no matter the outcome of that discussion, I don't think your comment in the test file here is doing any favor to future readers of this test. How does it help understand the test in any way?

Your comment is really a rationale for the change in the PR, not suitable documentation for the end result -- as such it should be part of the commit message, if anything.
___
*[pkg/sql/opt/exec/execbuilder/relational_builder.go, line 1198 at r6](https://reviewable.io/reviews/cockroachdb/cockroach/33339#-LUVLNJi5WP2NgMFkEBb:-LVOdLm91FFxb4NE0aKH:b-ich62n) ([raw file](https://github.com/cockroachdb/cockroach/blob/70e37d031d499630d757b1ef85a284bebb952e11/pkg/sql/opt/exec/execbuilder/relational_builder.go#L1198)):*
<details><summary><i>Previously, andy-kimball (Andy Kimball) wrote…</i></summary><blockquote>

I don't have more detail on the changes that will need to be made. The TODO is there to remind us to delve more deeply into this when we have time.
</blockquote></details>

acknowledged.


<!-- Sent from Reviewable.io -->

rytaft reviewed (approved):
<img class="emoji" title=":lgtm:" alt=":lgtm:" align="absmiddle" src="https://reviewable.io/lgtm.png" height="20" width="61"/>

> Reviewed 7 of 62 files at r7, 3 of 8 files at r8, 9 of 9 files at r10, 15 of 16 files at r11, 22 of 25 files at r12.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/33339#-:-LVPFE0Y01HH53tIzR2w:bnfp4nl)* status: :shipit: complete! 1 of 0 LGTMs obtained



<!-- Sent from Reviewable.io -->

rytaft reviewed (approved):
> Reviewed 4 of 62 files at r7, 3 of 25 files at r12.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/33339)* status: :shipit: complete! 1 of 0 LGTMs obtained



<!-- Sent from Reviewable.io -->

andy-kimball reviewed (commented):
OK, I added a release note, though I question whether we should publish it, given that this flag will be removed before our final 2.2 release.

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/33339#-:-LVP4Qu63GTooiPkg8Nc:b-4lwd61)* status: :shipit: complete! 0 of 0 LGTMs obtained (and 1 stale)

___
*[pkg/sql/logictest/testdata/logic_test/check_constraints, line 157 at r6](https://reviewable.io/reviews/cockroachdb/cockroach/33339#-LUVIyKQEFPqnkP1Z4X6:-LVPi68w5dbD5SMd5JrX:bb6un1j) ([raw file](https://github.com/cockroachdb/cockroach/blob/70e37d031d499630d757b1ef85a284bebb952e11/pkg/sql/logictest/testdata/logic_test/check_constraints#L157)):*
<details><summary><i>Previously, knz (kena) wrote…</i></summary><blockquote>

Yeah so the situation with release notes is that if you don't write something about this now, nobody will notice the change in behavior later when we switch the default. We're not creating release notes out of thin air about all the behavior that becomes different when a flag is switched. Someone has to write it, and I don't see any record of this behavior change anywhere for the future release note writer to pick up.

My recommendation would be for a release note that says "when the experimental flag is set, the behavior is ..."
</blockquote></details>

Done
___
*[pkg/sql/logictest/testdata/logic_test/computed, line 734 at r6](https://reviewable.io/reviews/cockroachdb/cockroach/33339#-LUVJDKdE3HSD7qHkQBt:-LVP5gzEE6LTl4wmV9w4:b-ibu16q) ([raw file](https://github.com/cockroachdb/cockroach/blob/70e37d031d499630d757b1ef85a284bebb952e11/pkg/sql/logictest/testdata/logic_test/computed#L734)):*
<details><summary><i>Previously, knz (kena) wrote…</i></summary><blockquote>

1) I read your response to my review comment as an additional requirement on how we're going to produce meaningful contextual strings. It's a non-trivial problem but it's solvable.

2) no matter the outcome of that discussion, I don't think your comment in the test file here is doing any favor to future readers of this test. How does it help understand the test in any way?

Your comment is really a rationale for the change in the PR, not suitable documentation for the end result -- as such it should be part of the commit message, if anything.
</blockquote></details>

I found it useful, but I don't really care much either way, so I removed it.


<!-- Sent from Reviewable.io -->

andy-kimball commented:
bors r+

andy-kimball commented:
bors r+

RaduBerinde reviewed (commented):
Sorry I didn't get a chance to review this in a timely manner. This is fantastic work! I am sure it was a huge effort to get all these details in place.

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/33339#-:-LV_ap2U5ZiPo1nATzxW:biz2agc)* status: :shipit: complete! 0 of 0 LGTMs obtained (and 1 stale)



<!-- Sent from Reviewable.io -->
